bftengine/src/preprocessor/PreProcessor.cpp

```diff
@@ -606,15 +606,19 @@ void PreProcessor::messageHandler<PreProcessReplyMsg>(MessageBase *msg) {
 
 void PreProcessor::msgProcessingLoop() {
   while (!msgLoopDone_) {
-    int count = 0;
     {
       std::unique_lock<std::mutex> l(msgLock_);
-      while (!msgLoopDone_ && msgs_.empty()) msgLoopSignal_.wait(l);
+      while (!msgLoopDone_ && msgs_.empty()) {
+        msgLoopSignal_.wait_until(l, chrono::steady_clock::now() + std::chrono::milliseconds(WAIT_TIMEOUT_MILLI));
+      }
       if (msgLoopDone_) break;
-      count = msgs_.read_available();
+      if (!msgs_.read_available()) {
+        LOG_FATAL(logger(), "queue empty on wakeup");
+        ConcordAssert(false);
```

Comment:
Couldn't it be some sporadic wake-up? I would ignore it and continue.

Reply:
the condition to reach this point is that if the msgs is empty - from the loop. If it waits, wakes up for some reason - but if the queue is empy, it should go sleep again. To be on the safe side you are right, but maybe then we will hide some bug?

Reply:
I actually don't think you need the check at all. The code is pretty clear. It will loop until there is a message or an exit.  The only way this would be a problem is if multiple threads were pulling off the queue. I don't believe that's the case. 

But you should probably change the loop above to use read_available() instead of empty() for thread safety as you mentioned in a prior comment.